### DISCUSSION TITLE
Timings

### DISCUSSION THREAD

**Discussion by lex27087** [Peter-van-Tol/LiteX-CNC#94] [Q&A]
Hi Peter. We've been puzzling over why for several days. +-200microseconds.Why are the timings so big? All the necessary optimizations have been made in the computer. On Intel network cards, the timings are smaller.But on realtek maps, these timings are already +-400 microseconds. A lot of settings have been made for the severed cards, which do not help in any way.A special draver 8168 is installed.With the litehm2 firmware, the timing picture is completely different. Thank you, I am waiting for your reply![IMG_20240226_172133_578.jpg](https://github.com/Peter-van-Tol/LiteX-CNC/assets/160588822/fddda0ea-b88b-4182-93f6-5360f00186f9)

![IMG_20240225_134436_334.jpg](https://github.com/Peter-van-Tol/LiteX-CNC/assets/160588822/51e96756-72e4-4fc1-ad6d-58643d043b61)

**Comment by lex27087** [comment]
Realtek ![IMG_20240225_182013_67.jpg](https://github.com/Peter-van-Tol/LiteX-CNC/assets/160588822/87cdbe68-02ca-48f7-9608-8174994cfce3)

![IMG_20240219_164447_104.jpg](https://github.com/Peter-van-Tol/LiteX-CNC/assets/160588822/308cb855-12df-4eb5-a394-1ba9607a0003)

**Comment by lex27087** [comment]
LiteHm2 
![IMG_20240228_221241_921.jpg](https://github.com/Peter-van-Tol/LiteX-CNC/assets/160588822/9f12d65e-4a0e-40ad-9c2f-32e67a62fd54)

![IMG_20240228_205725_723.jpg](https://github.com/Peter-van-Tol/LiteX-CNC/assets/160588822/f9cddd3e-073f-4843-a400-3eef0634335f)

**Reply by Peter-van-Tol** [reply]
I'm not familiar with `hm2_lite`, but one has to consider that:
- during the write phase, all commands of the modules are being interpreted. Especially the `stepgen` is considered to be heavy;
- during the read phase, there are actually two round trips to and from the card. First the addresses to be read are being sent to the card and secondly the actual data is read from the card.

Considering that the read in you images is longer then the write action (which is a-typical), my suspicion is that the Etherbone protocol is the cause of these long periods. You can try to run `ping xxx.xxx.xxx.xxx` (where `xxx.xxx.xxx.xxx` is the ip-address of your card) to get an indication on the latency of the card. With a good connection, this should be around 0.1 to 0.2 ms. If this period is higher, then the network card might not be a good candidate for this job. It is strongly recommended to have a dedicated connection, i.e. there should not be a router or switch between the PC and xthe FPGA.

The communication protocol of MESA (using UDP) is different an maybe more optimal in your case. However, I'm bound to the Etherbone-protocol.

SPI is slower then Ethernet, because there is no latency for the network, it is turns out to be faster.

**Reply by lex27087** [reply]
@Peter-van-Tol Peter. As it turned out, this is a glitch with such low timings. I noticed one thing! When the interface is running and it's worth pulling out and plugging in the cable. That reading time drops almost 3 times. As soon as you restart the interface. All the numbers become what they were before the cable was pulled out.

**Reply by lex27087** [reply]
![IMG_20240229_122019_834.jpg](https://github.com/Peter-van-Tol/LiteX-CNC/assets/160588822/df7bb80e-447f-4482-88a4-27cb748b4174)

**Reply by Peter-van-Tol** [reply]
So it is solved?

**Reply by lex27087** [reply]
@Peter-van-Tol Well, that's how timings fall if you pull out and stick in the cable. When the interface is running. I do not know, I decided to show you just.

**Reply by Peter-van-Tol** [reply]
Can you share:
- `ping xxx.xxx.xxx.xxx`;
- your config
- your HAL-file
- information on the system you use (i.e. type / version of Linux)

I wil try to replicate this.

**Reply by lex27087** [reply]
@Peter-van-Tol 
https://cloud.mail.ru/public/PH25/7YMDdFAt1

**Reply by lex27087** [reply]
Linuxcnc 2.9.2
Linuxcnc 2.10

**Comment by EhsanRasouli1371** [comment]
What is the maximum frequency that can generate and receive pulses?
